src/tools/tidy/src/ui_tests.rs

```diff
@@ -105,13 +161,30 @@ fn recursively_check_ui_tests<'issues>(
     check: &mut RunningCheck,
     path: &Path,
     allowed_issue_names: &'issues BTreeSet<&'issues str>,
-) -> BTreeSet<&'issues str> {
+) -> (BTreeSet<&'issues str>, BTreeSet<String>) {
     let mut remaining_issue_names: BTreeSet<&str> = allowed_issue_names.clone();
+    let mut cur_subdirs: BTreeSet<String> = BTreeSet::new();
 
     let (ui, ui_fulldeps) = (path.join("ui"), path.join("ui-fulldeps"));
     let paths = [ui.as_path(), ui_fulldeps.as_path()];
     crate::walk::walk_no_read(&paths, |_, _| false, &mut |entry| {
         let file_path = entry.path();
+        let subdirectory = format!(
+            "{}/",
+            file_path
+                .strip_prefix(path)
+                .unwrap()
+                .to_str()
+                .unwrap()
+                .split(std::path::MAIN_SEPARATOR_STR)
+                .take(2)
+                .collect::<Vec<_>>()
+                .join("/")
+        );
+        if !file_path.ends_with("ui/README.md") && !subdirectory.starts_with("ui-fulldeps/") {
+            cur_subdirs.insert(subdirectory);
+        }
```

Comment:
Doing this string manipulation on every file seems wasteful to me, can't we read all the directories at once (without recursion)?

Maybe with something like this:
```rust
let entries = std::fs::read_dir(ui.as_path())?;

let cur_subdirs: Vec<String> = entries
        .filter_map(|entry| entry.ok())
        .map(|entry| entry.path())
        .filter(|path| path.is_dir())
        .map(|dir_path| {
            let dir_path = dir_path.strip_prefix(path).unwrap();
            dir_path.to_string_lossy().replace(std::path::MAIN_SEPARATOR_STR, "/")
        })
        .collect();
```

Reply:
and also removed the trailing slashes from the directory strings in `ui_suites.txt`. Having the slashes at the end was making the sorting look a bit off.